### ISSUE TITLE
length-zero-no-unit doesn't warn in some cases.

### ISSUE THREAD

**Issue by XhmikosR** [stylelint/stylelint#2940]
Recently we switched the Bootstrap repo to stylelint. Things seem good so far :)

Now, one thing I noticed is that even though I have `length-zero-no-unit` set to `true`, we don't get any warning for the following case:

```scss
@mixin make-col-offset($size, $columns: $grid-columns) {
  margin-left: percentage($size / $columns);
}
```

`$size/$columns` can be 0 thus resulting to `margin-left: 0%;`.

I hope this is the right place to report the issue.

Thanks in advance!

EDIT: that's with stylelint 8.1.1.

**Comment by XhmikosR** [comment]
After reading the docs more carefully, I see that **length** refers to distance measurements only. 

Still if I change the mixin to return pixels I still don't get any warning.

```scss
@mixin make-col-offset($size, $columns: $grid-columns) {
  $num: $size / $columns;
  margin-left: if($num == 0, 0, round($num) + px);
}
```

**Comment by jeddy3** [comment]
@XhmikosR Thanks for creating the issue.

> Recently we switched the Bootstrap repo to stylelint. Things seem good so far :)

That's excellent news.

> Still if I change the mixin to return pixels I still don't get any warning.

Unfortunately, I'm struggling to understand your request. Can you help me? Am I right in thinking you're anticipating that stylelint will understand the _result_ of a SCSS `percentage` and `if` function?

> I hope this is the right place to report the issue.

Yes it is. It would be great if you could use the [issue template](https://github.com/stylelint/stylelint/blob/master/.github/ISSUE_TEMPLATE.md) for future issues as it helps us better understand your request.

**Comment by XhmikosR** [comment]
I was under the impression that stylelint is using a postcss parser and thus understands Sass. So, yeah, I'd expect stylelint to work for this case too.

BTW, I think the `0%` warning can be a separate issue since it affects everything. But I guess there must be some kind of logic behind that decision.

**Comment by jeddy3** [comment]
> I was under the impression that stylelint is using a postcss parser and thus understands Sass. So, yeah, I'd expect stylelint to work for this case too.

Oh, I see. Unfortunately, that's not how the PostCSS parser nor stylelint works. 

>  and thus understands Sass

PostCSS can parse SCSS into an Abstract Syntax Tree (AST) e.g. [here is](https://astexplorer.net/#/gist/d2440cc677802a6becb42812656abea7/53c1b5c3b5d9349a116ee3ba98edee860e8bb020) what your example mixin looks like when parsed by the default PostCSS parser. Notice how the `if` function is simply parsed as the `value` of the last `decl`. PostCSS doesn't understand that `if` and `percentage` are functions that can return a _length_ i.e. it cannot compile SCSS, but it can parse it into an AST. 

stylelint then traverses through the AST and conducts its static analysis on the nodes. 

I believe the PostCSS SCSS parser is very similar the default CSS one, but it can deal with constructs unique to SCSS, like `//` comments. Whereas the PostCSS default parser [will likely treat](https://astexplorer.net/#/gist/d2440cc677802a6becb42812656abea7/043228b02b1039c00a1da045b776e3643fb8ed95) `//` comments as a syntax error as they are not part of the CSS specification.

Does that make sense? Let me know if there are any bits that don't.

I believe that, despite not being able to compile SCSS, stylelint can be valuable to SCSS users. We've taken great care to make sure the built-in rules do not trip up on non-standard SCSS constructs like variables, maps, variable interpolation, `//` comments etc. This has then enabled plugin authors, see [stylelint-scss](https://github.com/kristerkari/stylelint-scss#list-of-rules), to create rules specific to these constructs.

> BTW, I think the 0% warning can be a separate issue since it affects everything. But I guess there must be some kind of logic behind that decision.

Can you please create a separate issue for that (using the template please), so that we can diagnose what's going on and help you out?

**Comment by XhmikosR** [comment]
@jeddy3: thanks for the detailed explanation!

Just to be clear this didn't throw with scss-lint either. I didn't think the issue a lot before reporting it but it makes sense now.

I'll try to open a new issue about zero percentages.
